packages/hir/src/lowering/types.candy

```diff
@@ -161,12 +161,26 @@ fun astInlineTypeToHirInlineType(
 }
 
 impl HirImpl {
-  fun constraints(): Map<HirParameterType, HirInlineType> {
+  fun constraints(context: QueryContext<List<CompilerError>>): Map<HirParameterType, HirInlineType> {
     // The type constraints of an impl. For example, `impl[T: Equals] Foo[T]: Equals` has the
     // constraints `T: Equals`.
 
-    // TODO(marcelgarus): Support constraints.
-    Map.empty<HirParameterType, HirInlineType>()
+    query<Map<HirParameterType, HirInlineType>, List<CompilerError>>(
+      context, "HirImpl.constraints", this as Equals & Hash, {
+
+      let constraints = MutableMap.empty<HirParameterType, HirInlineType>()
+
+      for parameter in this.ast(context).typeParameters {
+        if (parameter.upperBound is Some) {
+          constraints.set(
+            HirParameterType(this, parameter.name.value),
+            astInlineTypeToHirInlineType(context, parameter.upperBound.unwrap(), this),
+          )
+        }
+      }
```

Comment:
Type parameters are already lowered into the HIR: `hirImpl.typeParameters(context)` gives you a `List<HirParameterType>`. And getting the upper bound should probably be added as a query on `HirParameterType`.